docs/labs/format-strings.html

```diff
@@ -0,0 +1,136 @@
+<meta http-equiv="X-UA-Compatible" content="IE=edge">
+<meta name="viewport" content="width=device-width, initial-scale=1">
+<link rel="stylesheet" href="https://best.openssf.org/assets/css/style.css">
+<link rel="stylesheet" href="checker.css">
+<script src="js-yaml.min.js"></script>
+<script src="checker.js"></script>
+<link rel="license" href="https://creativecommons.org/licenses/by/4.0/">
+
+
+<script id="expected0" type="plain/text">
+    return '{format_string}, {event.level}'.format(format_string=user_input, event=event)
+</script>
+
+<!-- Full pattern of correct answer -->
+<script id="correct0" type="plain/text">
+\s*return\s'\{format_string\},\s\{event\.level\}'\.format\(format_string=user_input,\sevent=event\)\s*
+</script>
+
+<script id="info" type="application/yaml">
+---
+- absent: "user_input"
+  text: Make sure the user_input is included in the replacement fields passed to the format function
+  examples:
+  - - "'some format'.format(user_input=user_input)"
+  - - "'some format'.format(user_input)"
+- absent: "event"
+  text: Make sure the event is included in the replacement fields passed to the format function
+  examples:
+  - - "'some format'.format(user_input=user_input, event=event)"
+  - - "'some format'.format(user_input, event)"
+- absent: \{(0|format_string)\}
+  text: Make sure the braces for the format_string replacement field are in the format string
+  examples:
+  - - "'{event}'.format(format_string=user_input)"
+  - - "'{0}.format(user_input)"
+- absent: \{(1|event)\}
+  text: Make sure the braces for the event replacement field are in the format string
+  examples:
+  - - "'{format_string}, {event.level}'.format(format_string=user_input, event=event)"
+  - - "'{0}, {1.level}'.format(user_input, event)"
+- absent: \{(1\.level|event\.level)\}
+  text: Make sure the level attribute of the event replacement field is in the format string
+  examples:
+  - - "'{format_string}, {event.level}'.format(format_string=user_input, event=event)"
+  - - "'{0}, {1.level}'.format(user_input, event)"
+# debug: true
+</script>
+</head>
+<body>
+<!-- For GitHub Pages formatting: -->
+<div class="container-lg px-3 my-5 markdown-body">
+<h1>Lab Exercise Format Strings and Templates</h1>
+<p>
+This is a lab exercise on developing secure software.
+For more information, see the <a href="introduction.html" target="_blank">introduction to
+the labs</a>.
+
+<p>
+<h2>Task</h2>
+<p>
+<b>Practice using string templates in a secure way.</b>
+
+<p>
+<h2>Background</h2>
+<p>
+In this exercise, we'll adjust our string formatting so that it doesn't allow a user to control
+the <a href="https://docs.python.org/3/tutorial/inputoutput.html#the-string-format-method"><tt>
+format string</tt></a>. If a user can control the <tt>format string</tt> they can access
+variables which they shouldn't. Particularly if those variable's values can be returned to the user
+as output, it could lead to information disclosure beyond what was intended by the developer.
```

Comment:
I'd add something like this:

> In some cases the results can be even more serious, depending on the situation and programming language used. If a similar problem happened in C's printf function, it would often enable an attacker to completely control the program.